packages/lib/gpcircuits/src/list-membership.ts

```diff
@@ -0,0 +1,183 @@
+ * @param memberIndices the indices of the entry values (or tuples) each of which is a member of the corresponding list
+ * @param lists the membership lists of tuples
+ * @returns list of tuple indices of arity `params.tupleArity` representing the input tuples,
+ * numbers representing the indices of the entry values/entry value tuples which are members
+ *  of the lists, and the membership lists in hashed form.
+ * @throws RangeError if any of the indices are out of bounds.
+ * @throws TypeError if the list and index arrays are malformed.
+ */
+export function processLists(
+  params: ProtoPODGPCCircuitParams,
+  memberIndices: number[][],
+  lists: PODValue[][][]
+): {
+  tupleIndices: CircuitSignal[][];
+  memberIndex: CircuitSignal[];
+  membershipList: CircuitSignal[][];
+} {
+  let firstTupleIndex = params.maxEntries;
+  const unpaddedOutputObject: {
+    tupleIndices: CircuitSignal[][];
+    memberIndices: CircuitSignal[];
+    membershipLists: CircuitSignal[][];
+  } = {
+    tupleIndices: [],
+    memberIndices: [],
+    membershipLists: []
+  };
+
+  for (let i = 0; i < memberIndices.length; i++) {
+    // Process each of the lists
+    const processedList = processSingleList(
+      params,
+      firstTupleIndex,
+      memberIndices[i],
+      lists[i]
+    );
+
+    // Push tuple indices as bigints.
+    processedList.tupleIndices.forEach((indexTuple: number[]) =>
+      unpaddedOutputObject.tupleIndices.push(indexTuple.map(BigInt))
+    );
+    // Push member indices as bigints.
+    unpaddedOutputObject.memberIndices.push(BigInt(processedList.memberIndex));
+
+    // Push the hashed membership list.
+    unpaddedOutputObject.membershipLists.push(processedList.membershipList);
+
+    // Increment the first tuple index.
+    firstTupleIndex += processedList.tupleIndices.length;
+  }
+
+  // Pad and return
+  return {
+    tupleIndices: padArray(
+      unpaddedOutputObject.tupleIndices,
+      params.maxTuples,
+      // Pad with tuples of 0.
+      extendedSignalArray([], params.tupleArity)
+    ),
+    memberIndex: extendedSignalArray(
+      unpaddedOutputObject.memberIndices,
+      params.maxLists,
+      // Pad with -1 (mod p).
+      BABY_JUB_NEGATIVE_ONE
+    ),
+    membershipList: padArray(
+      unpaddedOutputObject.membershipLists,
+      params.maxLists,
+      // Pad with lists of zeroes.
+      extendedSignalArray([], params.maxListEntries)
+    )
+  };
+}
+
+/**
+ * Processes a single membership list together with the (multi-)index
+ * of the entry value (or entry value tuple) that ought to
+ * be a member of this list. This is done by means of appropriate
+ * applications of {@link hashTuple} and {@link computeTupleIndices}.
+ * If no tuples are involved, then `memberIndex` is a singleton and
+ * `list` a list of singletons.
+ *
+ * @param params parameters of the ProtoPODGPC the list is processed for
+ * @param firstTupleIndex the index of the first output tuple in the combined array of
```

Comment:
I think we clarified this below, but for reference: It will be `params.maxEntries` the first time it is called. The value will be greater on subsequent calls assuming tuples have been 'allocated'.